fastcrypto/src/secp256r1.rs

```diff
@@ -0,0 +1,492 @@
+}
+
+impl Verifier<Secp256r1Signature> for Secp256r1PublicKey {
+    fn verify(&self, msg: &[u8], signature: &Secp256r1Signature) -> Result<(), signature::Error> {
+        // TODO: Until we have a recovery id, we recover both candidates for public keys
+        let pks = signature
+            .recover(msg)
+            .map_err(|_| signature::Error::new())?;
+
+        if !pks.contains(self) {
+            return Err(signature::Error::new());
+        }
+
+        self.pubkey
+            .verify(msg, &signature.sig)
+            .map_err(|_| signature::Error::new())
+    }
+}
+
+impl Secp256r1PublicKey {}
+
+impl AsRef<[u8]> for Secp256r1PublicKey {
+    fn as_ref(&self) -> &[u8] {
+        self.bytes
+            .get_or_try_init::<_, eyre::Report>(|| Ok(self.pubkey.as_ref().to_bytes().into()))
+            .expect("OnceCell invariant violated")
+    }
+}
+
+impl ToFromBytes for Secp256r1PublicKey {
+    fn from_bytes(bytes: &[u8]) -> Result<Self, FastCryptoError> {
+        match ExternalPublicKey::try_from(bytes) {
+            Ok(pubkey) => Ok(Secp256r1PublicKey {
+                pubkey,
+                bytes: OnceCell::new(),
+            }),
+            Err(_) => Err(FastCryptoError::InvalidInput),
+        }
+    }
+}
+
+impl Default for Secp256r1PublicKey {
+    fn default() -> Self {
+        Secp256r1PublicKey::from_bytes(&[0u8; PUBLIC_KEY_SIZE]).unwrap()
+    }
+}
+
+impl Display for Secp256r1PublicKey {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        write!(f, "{}", Base64::encode(self.as_ref()))
+    }
+}
+
+// There is a strong requirement for this specific impl. in Fab benchmarks
+impl Serialize for Secp256r1PublicKey {
+    fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
+    where
+        S: serde::Serializer,
+    {
+        serializer.serialize_str(&self.encode_base64())
+    }
+}
+
+impl<'de> Deserialize<'de> for Secp256r1PublicKey {
+    fn deserialize<D>(deserializer: D) -> Result<Self, D::Error>
+    where
+        D: de::Deserializer<'de>,
+    {
+        let s = <String as serde::Deserialize>::deserialize(deserializer)?;
+        let value = Self::decode_base64(&s).map_err(|e| de::Error::custom(e.to_string()))?;
+        Ok(value)
+    }
+}
+
+impl<'a> From<&'a Secp256r1PrivateKey> for Secp256r1PublicKey {
+    fn from(secret: &'a Secp256r1PrivateKey) -> Self {
+        Secp256r1PublicKey {
+            pubkey: ExternalPublicKey::from(&secret.privkey),
+            bytes: OnceCell::new(),
+        }
+    }
+}
+
+impl SigningKey for Secp256r1PrivateKey {
+    type PubKey = Secp256r1PublicKey;
+    type Sig = Secp256r1Signature;
+    const LENGTH: usize = PRIVATE_KEY_SIZE;
+}
+
+impl ToFromBytes for Secp256r1PrivateKey {
+    fn from_bytes(bytes: &[u8]) -> Result<Self, FastCryptoError> {
+        match ExternalSecretKey::try_from(bytes) {
+            Ok(privkey) => Ok(Secp256r1PrivateKey {
+                privkey,
+                bytes: OnceCell::new(),
```

Comment:
Well, if the deserialization succeeded, we happen to have a copy of the bytes, maybe we could place them in the `OnceCell`?